deps: update c-ares to 1.18.1 (#40660)
nodejs/node: doc/guides/maintaining-c-ares.md

# Maintaining c-ares

Updates to the c-ares dependency involve the following steps:
1. Downloading the source archive for the new version.

> 1. Unpacking the source in a temporary workspace directory.
> 1. Removing the `test` directory (to save disk space).
> 1. Copying over the existing `.gitignore`, pre-generated `config` directory and

Reviewer: ```suggestion
2. Unpacking the source in a temporary workspace directory.
3. Removing the `test` directory (to save disk space).
4. Copying over the existing `.gitignore`, pre-generated `config` directory and
```

Reviewer 2: I don't understand why we want to do that. It's a feature of markdown to write ordered lists only with ones, and I like it because it allows to add an item anywhere in the list without having te renumber everything after it.

Author: I've run the formatter and committed the results (plus the comma change and reworded the bit about LICENSE updates). I kind of agree with @targos re. the GitHub flavored markdown feature but I'm not going to fight the formatter.

Reviewer: @targos @richardlau This is a bit of a spaces vs. tabs kind of thing I suppose, but for what it's worth, I think it's better to format it this way so that the raw markdown is more easily readable. It's supposed to be one of the advantages of markdown that it is useable/understandable in its raw format. It makes things slightly harder for the author, but I think when the differences are small, we should favor the reader. (That said, I haven't looked to see if there is a configuration option in remark to change this behavior. There might be.)

Reviewer: (And I also don't actually feel strongly about it. We do more writing markdown than reading the raw markdown probably, so... I'm just going to submit to the formatter... ¯\\_(ツ)_/¯ )

   `cares.gyp` files.
1. Replacing the existing `deps/cares` with the workspace directory.
1. Modifying the `cares.gyp` file for file additions/deletions.

> 1. Rebuilding the main Node.js `LICENSE`.

Reviewer 3: Not sure if this needs to be formatted @Trott.

Reviewer: I added suggestions from the output of `node tools/lint-md/lint-md.src.mjs --format doc/guides/maintaining-c-ares.md`. (The comma suggestion is not from the tool, of course. But the others all are.)

Reviewer 3: Thank you @Trott!

> 1. Replacing the existing `deps/cares` with the workspace directory.
> 1. Modifying the `cares.gyp` file for file additions/deletions.
> 1. Rebuilding the main Node.js `LICENSE`.

Reviewer: ```suggestion
5. Replacing the existing `deps/cares` with the workspace directory.
6. Modifying the `cares.gyp` file for file additions/deletions.
7. Rebuilding the main Node.js `LICENSE`.
```

## Running the update script

The `tools/update-cares.sh` script automates the update of the c-ares source
files, preserving the existing files added by Node.js.

In the following examples, `x.y.z` should match the c-ares version to update to.

```console
./tools/update-cares.sh x.y.z
```

> e.g.

Reviewer: ```suggestion
e.g.

```

```console
./tools/update-cares.sh 1.18.1
```

## Check that Node.js still builds and tests

It may be necessary to update `deps/cares/cares.gyp` if any significant changes
have occurred upstream.

## Rebuild the main Node.js license

Run the `tools/license-builder.sh` script to rebuild the main Node.js `LICENSE`
file. This may result in no changes if c-ares' license has not changed.

```console
./tools/license-builder.sh
```

> If the `LICENSE` is updated for changes other than this c-ares update those

Reviewer: ```suggestion
If the `LICENSE` is updated for changes other than this c-ares update, those
```

> should be done in a separate pull request first.

Reviewer 3: I had to read this twice to understand the meaning. I believe "updated for changes other than this c-ares update" means changes to the `LICENSE` file that are caused by other dependencies? Maybe this can be clarified (but I don't have a simpler suggestion right now).

Author: I've reworded this to:
> If the updated `LICENSE` contains changes for other dependencies, those should
be done in a separate pull request first.

## Commit the changes

```console
git add -A deps/cares
```

> Add the rebuilt `LICENSE` if it has been updated.

Reviewer: ```suggestion
Add the rebuilt `LICENSE` if it has been updated.

```
